server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorFactory.java

```diff
@@ -59,18 +71,161 @@
     private final TermsAggregator.BucketCountThresholds bucketCountThresholds;
     private final boolean showTermDocCountError;
 
+    // TODO: Registration should happen on the actual aggregator classes
+    static void registerAggregators(ValuesSourceRegistry valuesSourceRegistry) {
+        valuesSourceRegistry.register(TermsAggregationBuilder.NAME, CoreValuesSourceType.BYTES, TermsAggregatorFactory.bytesSupplier(),
+            (fieldType, indexFieldData) -> fieldType.typeName().equals(KeywordFieldMapper.CONTENT_TYPE)
```

Comment:
I opted to use type name checks rather than `instanceof`, thinking it might be a bit more future-proof.  Happy to change if we don't think this is a good idea.

Reply:
Yeah, no super strong opinion...the strings just felt a little cleaner and future-proof'y (and I have an illogical distaste for `instanceof` checks).  I also liked that we can do things like explicitly whitelist what we mean a "numeric" is, rather than just what extends the numeric `IndexNumericFieldData`.   E.g. if a new plugin implements a new field type that has IndexNumericFieldData, core numeric aggs would automatically start using that.  Unsure if that's a bug or feature :)

But some types (like Range) would be a little more difficult due to how they are setup.  I didn't dig too much to see how hard it would be though.

Reply:
I think this is fine for this PR, but I want to think about it some more.  Might change it around before we go to master.  Honestly, I don't love passing in this lambda at all, doesn't feel like something people writing aggs should need to know about.  I'll think about it some more and we can touch base later in the week, but it's fine for this to merge as is.